icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -215,32 +224,84 @@ private String[] asStringArray(ICUResourceBundle resource) {
      * @param name The name to be formatted.
      * @return If true, use given-first order to format the name; if false, use surname-first order.
      */
-    private boolean nameIsGnFirst(PersonName name) {
-        // the name can declare its order-- check that first (it overrides any locale-based calculation)
-        if (name.getPreferredOrder() == PersonName.PreferredOrder.GIVEN_FIRST) {
-            return true;
-        } else if (name.getPreferredOrder() == PersonName.PreferredOrder.SURNAME_FIRST) {
-            return false;
+private boolean nameIsGnFirst(PersonName name) {
+    // the name can declare its order-- check that first (it overrides any locale-based calculation)
+    if (name.getPreferredOrder() == PersonName.PreferredOrder.GIVEN_FIRST) {
+        return true;
+    } else if (name.getPreferredOrder() == PersonName.PreferredOrder.SURNAME_FIRST) {
+        return false;
+    }
+
+    Set<String> namedLocales = new HashSet<>();
+    namedLocales.addAll(gnFirstLocales);
+    namedLocales.addAll(snFirstLocales);
+
+    // NOTE: Despite the similar names, getNameLocale(name) and name.getNameLocale() do different things:
+    // name.getNameLocale() just returns the contents of the name's nameLocale field, but getNameLocale(name)
+    // guesses a locale for the name (using the stated locale as a hint, if it exists), filling in a script
+    // based on the characters used in the actual name text.
+    // localeMatchingLanguageCode is there for a few instances (notably, zh_Hant) where both localeMatchingDerivedLocale
+    // and localeMatchingStatedLocale fall back to "und" because we're doing a cross-script match and we never try
+    // just the language code by itself.
+    Locale derivedLocale = getNameLocale(name);
+    String localeMatchingDerivedLocale = closestMatchingLocaleID(derivedLocale, namedLocales);
+    String localeMatchingStatedLocale = closestMatchingLocaleID(name.getNameLocale(), namedLocales);
+    String localeMatchingLanguageCode = closestMatchingLocaleID(new Locale(derivedLocale.getLanguage()), namedLocales);
```

Comment:
This is infuriatingly complicated compared to the algorithm in the spec, but I couldn't come up with a simpler way to achieve the same result.  Suggestions are welcome.